### ISSUE TITLE
sys-devel/gcc: shared objects are always generated

### ISSUE THREAD

**Issue by alaviss** [haikuports/haikuports#3319]
`gcc` will always pass `-shared` to the linker, telling it to generate a shared object for all circumstances:

https://github.com/haikuports/haikuports/blob/0337d6953dee78b0a8890d4632311fe368c0b437/sys-devel/gcc/patches/gcc-7.3.0_2018_05_01.patchset#L840-L841

This was done for technical reasons within Haiku, namely:
- Required for replicants (as told on IRC)
- Haiku's native TLS (thread-local storage) support for ELF requires "executables" to be shared objects https://github.com/haiku/haiku/commit/44c0c4d3d0d68c3c3db23858b1089d3cf51bdf16

This, however, comes with its own downsides:
- *All* symbols in a binary is exported, which causes them to override any symbols from `dlopen()`-ed library and can lead to bugs like the one in [Ticket #14531][0]. On other OS, this must be requested explicitly via `-rdynamic` to the compiler or `--export-dynamic` to the linker.
- `-static` on Haiku is a lie, `-shared` have precedence over that flag, so building anything with `-static` will not result in a static binary, but rather a shared one (with no error/warning message). We currently don't support static binaries as our syscall numbers aren't locked, so this should've generated an error message instead.
- Might slow down program start up, since every binaries expose all of their symbols, which, in theory, could cause `runtime_loader` to spend more time resolving them (should be negligible).

Since PIE has pretty much [the same schematic][1] as a shared object, but doesn't export all symbols and does generate error if you try `-static`, I've looked in to it as a replacement for the current implicit `-shared`. However, this causes a problem: most (if not all) programs that make use of TLS will randomly segfault.

I've also tried combining `-pie` with `--export-dynamic` and `--no-runtime-linker` to (hopefully) emulate `-shared` (which should fix TLS), but that generates unusable binaries instead (`cannot relocate: Bad data`).

Any help on this is appreciated!

[0]: https://dev.haiku-os.org/ticket/14531
[1]: http://www.openbsd.org/papers/nycbsdcon08-pie/mgp00007.html

**Comment by waddlesplash** [comment]
This really sounds like multiple bugs in runtime_loader, not in GCC, so why is this an issue here instead of at Haiku?

**Comment by pulkomandy** [comment]
On Thu, Nov 01, 2018 at 06:51:58PM +0000, alaviss wrote:
> - *All* symbols in a binary is exported, which causes them to override any symbols from `dlopen()`-ed library and can lead to bugs like the one in [Ticket #14531][0]. On other OS, this must be requested explicitly via `-rdynamic` to the compiler or `--export-dynamic` to the linker.

One should use -fvisibility=hidden to avoid exporting all symbols. It is
a good thing to do even for libraries (then you need to explicitly
export symbols that are part of the library public ABI).

**Comment by alaviss** [comment]
@pulkomandy 

> One should use -fvisibility=hidden to avoid exporting all symbols. It is a good thing to do even for libraries (then you need to explicitly export symbols that are part of the library public ABI).

Yes, but most software was written under the assumption that the compiler won't export all of the symbols in their binary, and/or that `-shared` will do the exporting implicitly.

@waddlesplash 

> This really sounds like multiple bugs in runtime_loader, not in GCC, so why is this an issue here instead of at Haiku?

> -static on Haiku is a lie, -shared have precedence over that flag, so building anything with -static will not result in a static binary, but rather a shared one (with no error/warning message). We currently don't support static binaries as our syscall numbers aren't locked, so this should've generated an error message instead.

That's a gcc bug :P

**Comment by alaviss** [comment]
> One should use -fvisibility=hidden to avoid exporting all symbols. It is a good thing to do even for libraries (then you need to explicitly export symbols that are part of the library public ABI).

Actually, we could make `hidden` the default like on Windows, but then that could cause some problems with naive library writers... not to mention the risk of breaking a lot of our ports.
